### ISSUE TITLE
not able to use packages in vendor directory in docker's repository

### ISSUE THREAD

**Issue by vigneshkm** [moby/moby#38764]
I am trying to create a container with docker's go api. I want to expose a port using `container.Config.ExposedPorts` in `ContainerCreate()` API. Below is the code:
```
package main

import (
        "fmt"
        "context"
        "github.com/docker/docker/api/types/container"
        "github.com/docker/docker/client"
        "github.com/docker/go-connections/nat"
)

func main() {

        ctx := context.Background()
        cli, err := client.NewClientWithOpts(client.WithVersion("1.38"))
        if err != nil {
                fmt.Println("Failed to get container envoronment", err)
        }   

        resp, err := cli.ContainerCreate(ctx, &container.Config{
                Image: "hyperledger/fabric-ca",
                Cmd:   []string{"/bin/sh", "-c", "fabric-ca-server start -b admin:adminpw"},
                Env: []string{"FABRIC_CA_HOME=/etc/hyperledger/fabric-ca-server",
                        "FABRIC_CA_SERVER_CA_NAME=ca.example.com"},
                ExposedPorts: nat.PortSet{"22/tcp":struct{}{},},
        }, nil, nil, "ca.example.com")


        if err != nil {
                fmt.Println(" failed to create container, err:", err)
        } else {
                fmt.Println(" Container ID :", resp.ID, "warning:", resp.Warnings, "err:", err)
        }   
}
```
when I compile I get the below error
```
vignesh@vignesh-ThinkPad-E470 ~/go-book/src/github.com/my_fabric $ go build asd.go 
asd.go:8:9: cannot find package "github.com/docker/go-connections/nat" in any of:
    /home/vignesh/go-book/src/github.com/my_fabric/vendor/github.com/docker/go-connections/nat (vendor tree)
    /usr/local/go/src/github.com/docker/go-connections/nat (from $GOROOT)
    /home/vignesh/go-book/src/github.com/docker/go-connections/nat (from $GOPATH)
```
As package "github.com/docker/go-connections/nat" is in a vendor directory at "github.com/docker/docker/vendor/github.com/docker/go-connections/nat", I then created a vendor directory in my working directory and copied contents of github.com/docker/docker/vendor/github.com/docker/go-connections/nat to github.com/my_fabric/vendor/go-connections/nat and used "github.com/my_fabric/go-connections/nat" in import rather than "github.com/docker/go-connections/nat". But I got the following error.
```
vignesh@vignesh-ThinkPad-E470 ~/go-book/src/github.com/my_fabric $ go build asd.go 
# command-line-arguments
./asd.go:25:29: cannot use "github.com/my_fabric/vendor/github.com/my_fabric/go-connections/nat".PortSet literal (type "github.com/my_fabric/vendor/github.com/my_fabric/go-connections/nat".PortSet) as type "github.com/docker/docker/vendor/github.com/docker/go-connections/nat".PortSet in field value
```
Basically I want to use packages which is in vendor directory in docker's repository. Kindly help :)

**Comment by thaJeztah** [comment]
You need to vendor all the dependencies that you're using in your project, but **strip** nested vendor directories so; if your project is at

```
~/go-book/src/github.com/my_fabric
```

Then all the dependencies, including docker's dependencies should be in that project's vendor, e.g. (just to illustrate; list of dependencies is not complete)


```
tree -d -L 3 ~/go-book/src/github.com/my_fabric/vendor
vendor
├── github.com
│   ├── Azure
│   │   └── go-ansiterm
│   ├── Microsoft
│   │   ├── go-winio
│   │   └── hcsshim
│   ├── containerd
│   │   ├── console
│   │   ├── containerd
│   │   ├── continuity
│   │   ├── fifo
│   │   └── typeurl
│   ├── docker
│   │   ├── distribution
│   │   ├── docker
│   │   ├── go-connections
│   │   ├── go-events
│   │   ├── go-metrics
│   │   ├── go-units
│   │   └── swarmkit
```

The `vendor/github.com/docker/docker/vendor` directory is _empty_, but you should make sure that the version of the dependencies you're vendoring in matches the versions that are used by docker (sometimes more recent versions work, but there's some dependencies that should have the exact same version).


Usually it works best to set a new `GOPATH` when building your project, so that you know it's only using dependencies from the `vendor` directory, and is not finding the dependencies somewhere in your gopath.

**Comment by vigneshkm** [comment]
@thaJeztah Thank you for your kind response.
As suggested by you, I copied all of dockers vendored dependencies to my projects vendor directory
```
vignesh@vignesh-ThinkPad-E470 ~/go-book/src/github.com/docker/docker/vendor $ ls
cloud.google.com  github.com  go.etcd.io  golang.org  google.golang.org  go.opencensus.io  gotest.tools

```
```
vignesh@vignesh-ThinkPad-E470 ~/go-book/src/github.com/my_fabric/vendor $ ls
cloud.google.com  github.com  go.etcd.io  golang.org  google.golang.org  go.opencensus.io  gotest.tools
```
With my usual gopath `/home/vignesh/go-book/`, I still get the below error.
```
vignesh@vignesh-ThinkPad-E470 ~/go-book/src/github.com/my_fabric $ go build asd.go 
# command-line-arguments
./asd.go:24:29: cannot use "github.com/my_fabric/vendor/github.com/docker/go-connections/nat".PortSet literal (type "github.com/my_fabric/vendor/github.com/docker/go-connections/nat".PortSet) as type "github.com/docker/docker/vendor/github.com/docker/go-connections/nat".PortSet in field value

```

Regarding setting a new GOPATH, As I still want to use packages like ```"github.com/docker/docker/api/types/container"``` from existing GOPATH, changing GOPATH to project directory ```"/home/vignesh/go-book/src/github.com/my_fabric"``` would make those packages inaccessible and compilation fails.

**Comment by vigneshkm** [comment]
@thaJeztah I created my vendor directory as shown by you in the tree. Copied docker's vendor directory at "vendor/github.com/docker", docker code at "vendor/github.com/docker/docker". I removed vendor directory from "vendor/github.com/docker/docker" to prevent nested vendor directory and I am now able to use the contents of vendored dependencies. Thank you for the suggestion :)
